lib/gems/pending/util/object_storage/miq_ftp_storage.rb

```diff
@@ -0,0 +1,92 @@
+require 'util/miq_ftp_lib'
+require 'util/miq_object_storage'
+require 'logger'
+
+class MiqFtpStorage < MiqObjectStorage
+  include MiqFtpLib
+
+  FTP_CHUNKSIZE = ::Net::FTP::DEFAULT_BLOCKSIZE
+
+  attr_reader :uri, :username, :password
+
+  class << self
+    attr_writer :instance_logger
+
+    def instance_logger
+      @instance_logger ||= $log || Logger.new(File::NULL)
+    end
+  end
+
+  def self.uri_scheme
+    "ftp".freeze
+  end
+
+  def initialize(settings)
+    super
+    @uri      = @settings[:uri]
+    @username = @settings[:username]
+    @password = @settings[:password]
+  end
+
+  # Override for connection handling
+  def add(*upload_args)
+    with_connection { super }
+  end
+
+  # Override for connection handling
+  def download(*upload_args)
+    with_connection { super }
+  end
+
+  # Specific version of Net::FTP#storbinary that doesn't use an existing local
+  # file, and only uploads a specific size (byte_count) from the input_file
+  def upload_single(dest_uri)
+    ftp.synchronize do
+      ftp.send(:with_binary, true) do
+        conn = ftp.send(:transfercmd, "STOR #{uri_to_relative(dest_uri)}")
+        IO.copy_stream(source_input, conn, byte_count)
+        conn.close
+        ftp.send(:voidresp)
+      end
+    end
+    dest_uri
+  rescue Errno::EPIPE
+    # EPIPE, in this case, means that the data connection was unexpectedly
+    # terminated.  Rather than just raising EPIPE to the caller, check the
+    # response on the control connection.  If getresp doesn't raise a more
+    # appropriate exception, re-raise the original exception.
+    getresp
```

Comment:
Where is this defined?

Reply:
Oops... yeah, this is part of `Net::FTP`, and it really should be `ftp.send(:getresp)`... pretty sure.

Yeah, screwed that one up, but don't have a good test around it currently.  Will see if I can work on up, but for now, I will confirm that `ftp.send(:getresp)` what makes sense since the code is a update to the internal `Net::FTP#storbinary`.